consensus/src/extension/mod.rs

```diff
@@ -1,189 +1,52 @@
-/// finalizer_tx endpoint.
+/// units that should be finalized, and uses the finalization handler to report that to the user.
 ///
 /// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/internals.html
 /// Section 5.4 for a discussion of this component.
-
-pub struct Service<H: Hasher> {
-    node_id: NodeIndex,
-    extender: Extender<H>,
-    electors: Receiver<ExtenderUnit<H>>,
-    finalizer_tx: Sender<Vec<H::Hash>>,
+pub struct Ordering<H: Hasher, D: Data, MK: MultiKeychain, FH: FinalizationHandler<D>> {
+    extender: Extender<DagUnit<H, D, MK>>,
+    finalization_handler: FH,
 }
 
-impl<H: Hasher> Service<H> {
-    pub fn new(
-        node_id: NodeIndex,
-        electors: Receiver<ExtenderUnit<H>>,
-        finalizer_tx: Sender<Vec<H::Hash>>,
-    ) -> Self {
+impl<H: Hasher, D: Data, MK: MultiKeychain, FH: FinalizationHandler<D>> Ordering<H, D, MK, FH> {
+    pub fn new(finalization_handler: FH) -> Self {
         let extender = Extender::new();
-        Service {
-            node_id,
+        Ordering {
             extender,
-            electors,
-            finalizer_tx,
-        }
-    }
-
-    pub async fn run(mut self, mut terminator: Terminator) {
-        let mut exiting = false;
-        let mut round = 0;
-        loop {
-            futures::select! {
-                v = self.electors.next() => match v {
-                    Some(u) => {
-                        debug!(target: LOG_TARGET, "{:?} New unit in Extender round {:?} creator {:?} hash {:?}.", self.node_id, u.round, u.creator, u.hash);
-                        for batch in self.extender.add_unit(u) {
-                            let head = *batch.last().expect("all batches are nonempty");
-                            if self.finalizer_tx.unbounded_send(batch).is_err() {
-                                warn!(target: LOG_TARGET, "{:?} Channel for batches should be open", self.node_id);
-                                exiting = true;
-                            }
-                            debug!(target: LOG_TARGET, "{:?} Finalized round {:?} with head {:?}.", self.node_id, round, head);
-                            round += 1;
-                        }
-                    },
-                    None => {
-                        warn!(target: LOG_TARGET, "{:?} Units for extender unexpectedly ended.", self.node_id);
-                        exiting = true;
-                    }
-                },
-                _ = terminator.get_exit().fuse() => {
-                    debug!(target: LOG_TARGET, "{:?} received exit signal.", self.node_id);
-                    exiting = true;
-                }
-            }
-            if exiting {
-                debug!(target: LOG_TARGET, "{:?} Extender decided to exit.", self.node_id);
-                terminator.terminate_sync().await;
-                break;
-            }
+            finalization_handler,
         }
     }
-}
-
-#[cfg(test)]
```

Comment:
The only thing this test was related to threading, which is no longer an issue here. The only thing we could additionally be testing here is whether the elected units get passed on to the handler (since all other aspects of the election are tested in submodules), which is kinda already covered by `testing::dag`. I could add a test here as well, but it would be quite annoying for little gain imo.